presto-main/src/main/java/com/facebook/presto/operator/StreamingAggregationOperator.java

```diff
@@ -0,0 +1,294 @@
+                }
+            }
+            previousRow = null;
+        }
+
+        int startPosition = 0;
+        while (true) {
+            int groupEnd = findGroupEnd(startPosition, page);
+            addRowsToAggregates(page, startPosition, groupEnd);
+
+            if (groupEnd < page.getPositionCount()) {
+                evaluateAggregatesToPageBuilder(page, startPosition);
+
+                if (pageBuilder.isFull()) {
+                    pendingInput = page.getRegion(groupEnd, page.getPositionCount() - groupEnd);
+                    return;
+                }
+
+                startPosition = groupEnd;
+            }
+            else {
+                previousRow = page.getRegion(page.getPositionCount() - 1, 1);
+                pendingInput = null;
+                return;
+            }
+        }
+    }
+
+    private void addRowsToAggregates(Page page, int startPosition, int endPosition)
+    {
+        for (Aggregator aggregator : aggregates) {
+            aggregator.processPage(page.getRegion(startPosition, endPosition - startPosition));
+        }
+    }
+
+    private void evaluateAggregatesToPageBuilder(Page page, int position)
+    {
+        pageBuilder.declarePosition();
+        for (int i = 0; i < groupByTypes.size(); i++) {
+            Block block = page.getBlock(groupByChannels.get(i));
+            Type type = groupByTypes.get(i);
+            type.appendTo(block, position, pageBuilder.getBlockBuilder(i));
+        }
+        int offset = groupByTypes.size();
+        for (int i = 0; i < aggregates.size(); i++) {
+            aggregates.get(i).evaluate(pageBuilder.getBlockBuilder(offset + i));
+        }
+
+        aggregates = setupAggregates(step, accumulatorFactories);
+    }
+
+    private int findGroupEnd(int startPosition, Page page)
+    {
+        for (int i = startPosition + 1; i < page.getPositionCount(); i++) {
+            if (!rowEqualsRow(startPosition, page, i, page)) {
+                return i;
+            }
+        }
+
+        return page.getPositionCount();
+    }
+
+    private boolean rowEqualsRow(int position, Page page, int otherPosition, Page otherPage)
+    {
+        for (int i = 0; i < groupByChannels.size(); i++) {
+            int channel = groupByChannels.get(i);
+            Block block = page.getBlock(channel);
+            Block otherBlock = otherPage.getBlock(channel);
+            if (!TypeUtils.positionEqualsPosition(groupByTypes.get(i), block, position, otherBlock, otherPosition)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        if (pageBuilder.isFull()) {
+            Page output = pageBuilder.build();
+            pageBuilder.reset();
+            return output;
+        }
+
+        if (pendingInput != null) {
+            processInput(pendingInput);
+            updateMemoryUsage();
+            return null;
+        }
+
+        if (finishing) {
```

Comment:
you could flush unfinished aggregates in `finish()` method.

If you follow my suggestion with `outputPages`, then this method would greatly simplify